File: backend/src/routers/tasks.py
from fastapi import APIRouter, Depends, Query, Path, Request
from typing import Optional
from src.core.dependencies import get_current_user
from src.services import TaskService
from src.models import (
    TaskType,
    TaskSubType,
    TaskStatus,
    TaskPriority,
    ProgressInfo,
)
from src.core.handler import router_handler
from config.logging import get_router_logger, get_trace_id
from src.routers.schemas.task import TaskReadResponseSchema, TaskPageResultResponseSchema
from src.core.idempotency import idempotent

router = APIRouter(prefix="/api/v1/tasks", tags=["Tasks"])
service = TaskService()
logger = get_router_logger("tasks")

# -------------------- Query --------------------

@router.get("/", response_model=TaskPageResultResponseSchema)
@router_handler(action="list_tasks")
async def list_tasks(
    request: Request,
    current_user=Depends(get_current_user),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    query: Optional[str] = Query(None, description="任务名/描述模糊搜索"),
    task_type: Optional[TaskType] = Query(None, description="任务主类型筛选"),
    sub_type: Optional[TaskSubType] = Query(None, description="任务子类型筛选"),
    status: Optional[TaskStatus] = Query(None, description="状态筛选"),
    priority: Optional[TaskPriority] = Query(None, description="优先级筛选"),
    user_id: Optional[str] = Query(None, description="按用户筛选（仅管理员）"),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    return await service.list_tasks(
        context=context,
        page=page,
        size=size,
        query=query,
        task_type=task_type,
        sub_type=sub_type,
        status=status,
        priority=priority,
        user_id=user_id,
    )


@router.get("/{task_id}", response_model=TaskReadResponseSchema)
@router_handler(action="get_task_detail")
async def get_task_detail(
    request: Request,
    task_id: str = Path(..., description="任务ID"),
    current_user=Depends(get_current_user),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    return await service.get_task_detail(task_id, context=context)


@router.get("/{task_id}/progress", response_model=ProgressInfo)
@router_handler(action="get_task_progress")
async def get_task_progress(
    request: Request,
    task_id: str = Path(..., description="任务ID"),
    current_user=Depends(get_current_user),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    return await service.get_task_progress(task_id, context=context)

# -------------------- Update --------------------

@router.patch("/{task_id}/cancel", response_model=TaskReadResponseSchema)
@router_handler(action="cancel_task")
async def cancel_task(
    request: Request,
    task_id: str = Path(..., description="任务ID"),
    current_user=Depends(get_current_user),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    return await service.cancel_task(task_id, context=context)


@router.post("/{task_id}/retry", response_model=dict)
@router_handler(action="retry_task")
@idempotent()
async def retry_task(
    request: Request,
    task_id: str = Path(..., description="任务ID"),
    current_user=Depends(get_current_user),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    model_result, task_info = await service.retry_task(task_id, context=context)
    return {
        "data": model_result.model_dump(),
        "task": task_info,
    }


@router.delete("/{task_id}", response_model=bool)
@router_handler(action="delete_task")
async def delete_task(
    request: Request,
    task_id: str = Path(..., description="任务ID"),
    current_user=Depends(get_current_user),
):
    context = {
        "current_user": current_user,
        "trace_id": get_trace_id(),
        "request": {"path": str(request.url.path), "method": request.method},
    }
    return await service.delete_task(task_id, context=context)

File: frontend/src/api/tasks.ts
import { http } from "./http";
import type {
  TaskType,
  TaskSubType,
  TaskStatus,
  TaskPriority,
  TaskRead,
  TaskPageResult,
  ProgressInfo,
} from "@/types/task";

/**
 * 列表：GET /api/v1/tasks/
 */
export interface ListTasksParams {
  page?: number;
  size?: number;
  query?: string;
  task_type?: TaskType;
  sub_type?: TaskSubType;
  status?: TaskStatus;
  priority?: TaskPriority;
  user_id?: string;
}

export async function listTasks(
  token: string,
  params: ListTasksParams = {},
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<TaskPageResult> {
  const qp = new URLSearchParams();
  if (params.page !== undefined) qp.set("page", String(params.page));
  if (params.size !== undefined) qp.set("size", String(params.size));
  if (params.query) qp.set("query", params.query);
  if (params.task_type !== undefined && params.task_type !== null) {
    qp.set("task_type", params.task_type as unknown as string);
  }
  if (params.sub_type !== undefined && params.sub_type !== null) {
    qp.set("sub_type", params.sub_type as unknown as string);
  }
  if (params.status !== undefined && params.status !== null) {
    qp.set("status", params.status as unknown as string);
  }
  if (params.priority !== undefined && params.priority !== null) {
    qp.set("priority", String(params.priority));
  }
  if (params.user_id) qp.set("user_id", params.user_id);

  return http.get<TaskPageResult>("/api/v1/tasks/", {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

/**
 * 详情：GET /api/v1/tasks/{task_id}
 */
export async function getTaskDetail(
  token: string,
  taskId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<TaskRead> {
  return http.get<TaskRead>(`/api/v1/tasks/${taskId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 取消任务：PATCH /api/v1/tasks/{task_id}/cancel
 */
export async function cancelTask(
  token: string,
  taskId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<TaskRead> {
  return http.patch<TaskRead>(`/api/v1/tasks/${taskId}/cancel`, undefined, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 获取进度：GET /api/v1/tasks/{task_id}/progress
 */
export async function getTaskProgress(
  token: string,
  taskId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<ProgressInfo> {
  return http.get<ProgressInfo>(`/api/v1/tasks/${taskId}/progress`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 重试任务：POST /api/v1/tasks/{task_id}/retry
 */
interface RetryTaskResponse {
  data: TaskRead;
  task: { task_id: string; task_name: string };
}

export async function retryTask(
  token: string,
  taskId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<TaskRead> {
  const res = await http.post<RetryTaskResponse>(`/api/v1/tasks/${taskId}/retry`, undefined, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return res.data;
}

/**
 * 删除任务：DELETE /api/v1/tasks/{task_id}
 */
export async function deleteTask(
  token: string,
  taskId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<boolean> {
  return http.delete<boolean>(`/api/v1/tasks/${taskId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}